File: web/src/lib/sleep.ts
// Sleep quality — DERIVED figures, not raw HealthKit data.
//
// ⚠️ There is no HealthKit field called "sleep quality" (docs/23 §5). What exists
// is the series of `sleepAnalysis` category segments; total sleep time, the stage
// proportions, the efficiency and the number of awakenings are all DERIVED from
// it. So they are computed here, from the segments `/v1/sleep` returns.
//
// ⚠️ And the segments OVERLAP each other. `inBed` wraps the whole night, and
// since a second source (the watch beside the phone) began writing stages too,
// two sources describe the same night at their own boundaries. Adding up the raw
// lengths therefore gives roughly one and a half times the sleep time — which is
// why everything below is computed from the FLATTENED slices, never from the raw
// segments.
//
// The rules are the same three the phone (`HelsaKit/.../SleepAnalysis.swift`) and
// the backend (`backend/internal/sleep`) follow, and they have to stay the same:
// a night must not be one length on the dashboard and another on the phone.

import type { SleepSegment } from '../api/types'

/**
 * A gap longer than this starts a new sleep session.
 *
 * Grouping by calendar day is not an option: sleep runs across midnight, so the
 * segments before and after it would land on two different days, and every "day"
 * would hold the end of one night plus the beginning of the next — with roughly 24
 * hours of time in bed and a meaningless efficiency.
 */
const SESSION_GAP_MIN = 180

/** The normalized stages. The display names of these live in `i18n/*.ts`. */
export type Stage = 'deep' | 'rem' | 'core' | 'asleep' | 'awake' | 'inBed'

/**
 * Conflict resolution for overlapping stages — the SMALLER number wins.
 *
 * `awake` beats every sleep stage: if one source saw wakefulness, claiming deep
 * sleep in its place would be wrong in the more flattering direction, and
 * overestimating sleep is the worse error. `inBed` loses to everything, because
 * it is an envelope and not a measurement.
 */
const PRIORITY: Record<Stage, number> = {
  awake: 0,
  deep: 1,
  rem: 2,
  core: 3,
  asleep: 4,
  inBed: 5,
}

const ASLEEP: Record<Stage, boolean> = {
  deep: true,
  rem: true,
  core: true,
  asleep: true,
  awake: false,
  inBed: false,
}

/**
 * The raw `stage` string of `/v1/sleep` → a normalized stage.
 *
 * The database holds the same stage under two spellings: the HealthKit-flavoured
 * `asleepCore`/`asleepDeep`/`asleepREM` next to the short legacy
 * `core`/`deep`/`rem`. Anything unrecognized comes back as `null` — it stays
 * visible in the raw list below the night, but it does not silently become sleep.
 */
export function normalizeStage(raw?: string): Stage | null {
  switch ((raw ?? '').toLowerCase()) {
    case 'asleepdeep':
    case 'deep':
      return 'deep'
    case 'asleeprem':
    case 'rem':
      return 'rem'
    case 'asleepcore':
    case 'core':
    case 'asleeplight':
    case 'light':
      return 'core'
    case 'asleep':
    case 'asleepunspecified':
    case 'unspecified':
      return 'asleep'
    case 'awake':
    case 'awakened':
      return 'awake'
    case 'inbed':
    case 'in_bed':
      return 'inBed'
    default:
      return null
  }
}

/** A slice of the flattened timeline: disjoint, in milliseconds since the epoch. */
export type Slice = { start: number; end: number; stage: Stage }

export type Night = {
  /** The day of waking (local time, ISO date) — this identifies the night. */
  key: string
  /** The start of the session (ISO) — a unique identifier and the sort key. */
  startedAt: string
  /** The RAW segments, as they arrived — the "raw data" list is built from these. */
  segments: SleepSegment[]
  /** The overlap-free timeline. Every derived figure below comes from this. */
  slices: Slice[]
  /** How much the raw segments covered each other — the difference between the
   * naive sum and the truth. */
  overlapMin: number
  /** Minutes spent asleep: the UNION of the sleep slices, so an overlap counts once. */
  asleepMin: number
  /** The time-in-bed window: from the start of the first segment to the end of the last. */
  inBedMin: number
  /** time asleep / time in bed — 0..1, or null when it makes no sense. */
  efficiency: number | null
  /** Awakenings: `awake` slices BETWEEN falling asleep and the final wake-up. */
  awakenings: number
  /** Stage → minutes, from the flattened slices. */
  stages: Record<string, number>
  onset?: string
  wakeUp?: string
}

export function segmentMinutes(s: SleepSegment): number {
  if (!s.started_at || !s.ended_at) return 0
  return (new Date(s.ended_at).getTime() - new Date(s.started_at).getTime()) / 60000
}

export function sliceMinutes(s: Slice): number {
  return (s.end - s.start) / 60000
}

/** The local date in ISO form — the calendar day is the one on the user's clock,
 * not the UTC one. */
function localDate(iso: string): string {
  const d = new Date(iso)
  return `${d.getFullYear()}-${String(d.getMonth() + 1).padStart(2, '0')}-${String(d.getDate()).padStart(2, '0')}`
}

/**
 * Overlapping segments → DISJOINT slices.
 *
 * Every segment boundary is a cut point; between two neighbouring cut points
 * exactly one stage wins (see `PRIORITY`), and neighbouring slices of the same
 * stage are merged, so the hypnogram does not fall apart at invisible cuts.
 */
export function flatten(segments: SleepSegment[]): { slices: Slice[]; overlapMin: number } {
  const spans = segments
    .map((s) => ({
      start: s.started_at ? new Date(s.started_at).getTime() : NaN,
      end: s.ended_at ? new Date(s.ended_at).getTime() : NaN,
      stage: normalizeStage(s.stage),
    }))
    .filter((s): s is Slice => s.stage != null && s.end > s.start)

  if (spans.length === 0) return { slices: [], overlapMin: 0 }

  const cuts = [...new Set(spans.flatMap((s) => [s.start, s.end]))].sort((a, b) => a - b)
  const slices: Slice[] = []
  for (let i = 0; i + 1 < cuts.length; i++) {
    const [from, to] = [cuts[i], cuts[i + 1]]
    const covering = spans.filter((s) => s.start <= from && s.end >= to)
    if (covering.length === 0) continue // a gap: no stage reaches here
    const winner = covering.reduce((best, s) =>
      PRIORITY[s.stage] < PRIORITY[best.stage] ? s : best,
    ).stage

    const last = slices.at(-1)
    if (last && last.stage === winner && last.end === from) last.end = to
    else slices.push({ start: from, end: to, stage: winner })
  }

  const raw = spans.reduce((sum, s) => sum + (s.end - s.start), 0)
  const union = slices.reduce((sum, s) => sum + (s.end - s.start), 0)
  return { slices, overlapMin: Math.max(0, raw - union) / 60000 }
}

/**
 * Threads the segments into sleep sessions (nights): a continuous run of segments
 * is one night, and a gap longer than `SESSION_GAP_MIN` closes it. A daytime nap
 * therefore shows up as its own entry, instead of bleeding into the previous
 * night.
 */
export function groupByNight(segments: SleepSegment[], gapMin = SESSION_GAP_MIN): Night[] {
  const sorted = segments
    .filter((s) => s.started_at && s.ended_at)
    .sort((a, b) => a.started_at!.localeCompare(b.started_at!))

  const sessions: SleepSegment[][] = []
  let current: SleepSegment[] = []
  let lastEnd = 0
  for (const s of sorted) {
    const start = new Date(s.started_at!).getTime()
    const end = new Date(s.ended_at!).getTime()
    if (current.length > 0 && start - lastEnd > gapMin * 60_000) {
      sessions.push(current)
      current = []
      lastEnd = 0
    }
    current.push(s)
    lastEnd = Math.max(lastEnd, end)
  }
  if (current.length > 0) sessions.push(current)

  return sessions.map(analyseNight).sort((a, b) => b.startedAt.localeCompare(a.startedAt))
}

function analyseNight(input: SleepSegment[]): Night {
  const segments = [...input].sort((a, b) =>
    (a.started_at ?? '').localeCompare(b.started_at ?? ''),
  )
  const startedAt = segments[0]?.started_at ?? ''
  const lastEnd = segments
    .map((s) => s.ended_at)
    .filter((x): x is string => !!x)
    .sort()
    .at(-1)
  const key = lastEnd ? localDate(lastEnd) : localDate(startedAt)

  const { slices, overlapMin } = flatten(segments)

  const stages: Record<string, number> = {}
  let asleepMin = 0
  for (const s of slices) {
    const m = sliceMinutes(s)
    stages[s.stage] = (stages[s.stage] ?? 0) + m
    if (ASLEEP[s.stage]) asleepMin += m
  }

  // Time in bed is the whole spanning interval, not the sum of the segments: the
  // gap between two segments is time in bed too, and that is precisely what drags
  // efficiency down.
  const starts = segments.map((s) => s.started_at).filter((x): x is string => !!x)
  const ends = segments.map((s) => s.ended_at).filter((x): x is string => !!x)
  const inBedMin =
    starts.length && ends.length
      ? (Math.max(...ends.map((e) => new Date(e).getTime())) -
          Math.min(...starts.map((s) => new Date(s).getTime()))) /
        60000
      : 0

  const asleepSlices = slices.filter((s) => ASLEEP[s.stage])
  const onset = asleepSlices[0] ? new Date(asleepSlices[0].start).toISOString() : undefined
  const wakeUp = asleepSlices.at(-1) ? new Date(asleepSlices.at(-1)!.end).toISOString() : undefined

  // Only wakefulness AFTER falling asleep and BEFORE the final wake-up counts as an
  // awakening — tossing and turning before falling asleep does not.
  const awakenings =
    asleepSlices.length > 0
      ? slices.filter(
          (s) =>
            s.stage === 'awake' &&
            s.start >= asleepSlices[0].start &&
            s.end <= asleepSlices.at(-1)!.end,
        ).length
      : 0

  return {
    key,
    startedAt,
    segments,
    slices,
    overlapMin,
    asleepMin,
    inBedMin,
    efficiency: inBedMin > 0 ? Math.min(asleepMin / inBedMin, 1) : null,
    awakenings,
    stages,
    onset,
    wakeUp,
  }
}

export type SleepAverages = {
  nights: number
  asleepMin: number | null
  efficiency: number | null
  awakenings: number | null
  /** The share of deep + REM within total sleep time. */
  deepRemShare: number | null
}

export function averages(nights: Night[]): SleepAverages {
  if (nights.length === 0) {
    return { nights: 0, asleepMin: null, efficiency: null, awakenings: null, deepRemShare: null }
  }
  const mean = (xs: number[]) => (xs.length ? xs.reduce((a, b) => a + b, 0) / xs.length : null)
  const totalAsleep = nights.reduce((s, n) => s + n.asleepMin, 0)
  const deepRem = nights.reduce((s, n) => s + (n.stages.deep ?? 0) + (n.stages.rem ?? 0), 0)
  return {
    nights: nights.length,
    asleepMin: mean(nights.map((n) => n.asleepMin)),
    efficiency: mean(
      nights.map((n) => n.efficiency).filter((x): x is number => x != null),
    ),
    awakenings: mean(nights.map((n) => n.awakenings)),
    deepRemShare: totalAsleep > 0 ? deepRem / totalAsleep : null,
  }
}

File: backend/internal/sleep/sleep.go
// Package sleep turns the raw HealthKit sleep segments into an OVERLAP-FREE
// timeline, and answers the only question the rest of the backend actually asks
// of them: how long was I asleep?
//
// ⚠️ Why this cannot be a `sum(ended_at - started_at)` in SQL. The segments
// overlap each other, in two ways at once:
//
//   - `inBed` is not a measurement but an ENVELOPE: it wraps the whole night, so
//     adding it to the stages counts every minute twice.
//   - Since a second source (the watch next to the phone) started writing stages
//     as well, two sources describe the SAME night, in their own words, at their
//     own boundaries.
//
// Summing the raw lengths therefore reports roughly one and a half times the
// real sleep — a confident, believable lie, of the kind that is worse than
// silence. Every caller that wants a duration must go through Resolve or
// Nights.
//
// The semantics here deliberately mirror the phone's
// (`HelsaKit/Health/UI/SleepAnalysis.swift`): the same night has to come out the
// same length on the phone, on the web and in an insight. Two implementations of
// one rule is one implementation too many, and the two ends drifting apart is
// how the same rule fires on the phone and stays silent on the web.
package sleep

import (
	"sort"
	"time"
)

// Stage is a NORMALIZED sleep stage.
//
// Normalizing is not tidiness. In the real database the HealthKit-flavoured
// `asleepCore`/`asleepDeep`/`asleepREM` names live side by side with the short
// legacy `core`/`deep`/`rem` ones — for the same nights, overlapping each other.
// A query that only knows one of the two spellings silently drops half the data
// and shows LESS sleep than what was recorded, without anything looking broken.
type Stage string

const (
	Deep Stage = "deep"
	REM  Stage = "rem"
	Core Stage = "core"
	// Unspecified is HealthKit's `asleepUnspecified`: sleep, only the source did
	// not break it into stages. Treating it as zero (as the first version of the
	// insights query did) throws away whole nights recorded by a source that does
	// not do staging.
	Unspecified Stage = "asleepUnspecified"
	Awake       Stage = "awake"
	InBed       Stage = "inBed"
)

// IsAsleep: was this time actually spent asleep? `inBed` is NOT — it is the
// envelope between going to bed and getting up, and it overlaps every other
// stage.
func (s Stage) IsAsleep() bool {
	switch s {
	case Deep, REM, Core, Unspecified:
		return true
	default:
		return false
	}
}

// priority resolves a conflict between overlapping stages — the SMALLER number
// wins. Two rules stand behind the order:
//
//  1. **`awake` beats everyone.** If one source says I was awake, claiming deep
//     sleep in its place would be the stronger claim in the more unpleasant
//     direction. Overestimating sleep is the worse error.
//  2. **`inBed` loses to everyone.** It is an envelope, not a measurement: if
//     anyone says anything more specific for a moment, that wins.
//
// Among the sleep stages the more specific classification goes first. That part
// is ARBITRARY when two sources contradict each other — but it does not move the
// total (that is the union, see Night.Asleep), only the split between stages.
func (s Stage) priority() int {
	switch s {
	case Awake:
		return 0
	case Deep:
		return 1
	case REM:
		return 2
	case Core:
		return 3
	case Unspecified:
		return 4
	case InBed:
		return 5
	default:
		return 6
	}
}

// Normalize resolves a raw `stage` string. The second return value is false for
// a name we do not know — the caller must decide what to do with it, but it must
// not be counted as sleep: a new HealthKit stage silently counted as sleep is
// exactly the "confidently wrong upwards" answer this package exists to prevent.
func Normalize(raw string) (Stage, bool) {
	switch lower(raw) {
	case "asleepdeep", "deep":
		return Deep, true
	case "asleeprem", "rem":
		return REM, true
	case "asleepcore", "core", "asleeplight", "light":
		return Core, true
	case "asleep", "asleepunspecified", "unspecified":
		return Unspecified, true
	case "awake", "awakened":
		return Awake, true
	case "inbed", "in_bed":
		return InBed, true
	}
	return "", false
}

// lower is an ASCII-only strings.ToLower: the stage names are HealthKit
// identifiers, so there is nothing here for Unicode case folding to do.
func lower(s string) string {
	b := []byte(s)
	for i := range b {
		if b[i] >= 'A' && b[i] <= 'Z' {
			b[i] += 'a' - 'A'
		}
	}
	return string(b)
}

// Span is one stage over an interval. The interval is half-open, [Start, End).
type Span struct {
	Start time.Time
	End   time.Time
	Stage Stage
}

// Duration is never negative: a reversed span (bad data) is 0 long, it does not
// subtract from the night.
func (s Span) Duration() time.Duration {
	if !s.End.After(s.Start) {
		return 0
	}
	return s.End.Sub(s.Start)
}

// Raw is a segment as it comes out of the database: the stage is still a
// free-text string.
type Raw struct {
	Start time.Time
	End   time.Time
	Stage string
}

// Spans normalizes the raw rows and sorts them by start time. The unrecognized
// stage names come back SEPARATELY rather than disappearing: a stage we cannot
// place is a gap in the picture, and the caller gets to say so out loud.
//
// Zero-length and reversed segments are dropped here — they carry no time, and
// they would only add cut points to Resolve.
func Spans(rows []Raw) (spans []Span, unknown []string) {
	seen := map[string]bool{}
	for _, r := range rows {
		stage, ok := Normalize(r.Stage)
		if !ok {
			if !seen[r.Stage] {
				seen[r.Stage] = true
				unknown = append(unknown, r.Stage)
			}
			continue
		}
		if !r.End.After(r.Start) {
			continue
		}
		spans = append(spans, Span{Start: r.Start, End: r.End, Stage: stage})
	}
	sort.Slice(spans, func(i, j int) bool { return spans[i].Start.Before(spans[j].Start) })
	sort.Strings(unknown)
	return spans, unknown
}

// Resolve flattens overlapping spans into DISJOINT slices, and reports how much
// raw length the overlap had duplicated.
//
// Every stage boundary is a cut point; between two neighbouring cut points
// exactly one stage wins (see Stage.priority), and neighbouring slices of the
// same stage are merged so that the timeline does not fall apart at invisible
// cuts.
//
// The `overlap` is not a diagnostic curiosity: it is the difference between what
// the naive sum would have said and what we say, and it is the number to reach
// for when someone asks why the total is not the sum of the stages.
func Resolve(spans []Span) (slices []Span, overlap time.Duration) {
	if len(spans) == 0 {
		return nil, 0
	}

	cuts := cutPoints(spans)
	if len(cuts) < 2 {
		return nil, 0
	}

	var raw time.Duration
	for _, s := range spans {
		raw += s.Duration()
	}

	for i := 0; i+1 < len(cuts); i++ {
		from, to := cuts[i], cuts[i+1]
		winner, ok := covering(spans, from, to)
		if !ok {
			continue // a gap: not a single stage reaches here
		}
		if n := len(slices); n > 0 && slices[n-1].Stage == winner && slices[n-1].End.Equal(from) {
			slices[n-1].End = to
			continue
		}
		slices = append(slices, Span{Start: from, End: to, Stage: winner})
	}

	var union time.Duration
	for _, s := range slices {
		union += s.Duration()
	}
	if raw > union {
		overlap = raw - union
	}
	return slices, overlap
}

// cutPoints collects the deduplicated, ordered boundaries of the spans.
func cutPoints(spans []Span) []time.Time {
	seen := make(map[int64]time.Time, 2*len(spans))
	for _, s := range spans {
		seen[s.Start.UnixNano()] = s.Start
		seen[s.End.UnixNano()] = s.End
	}
	out := make([]time.Time, 0, len(seen))
	for _, t := range seen {
		out = append(out, t)
	}
	sort.Slice(out, func(i, j int) bool { return out[i].Before(out[j]) })
	return out
}

// covering picks the winning stage over the [from, to) interval: of the spans
// that cover it whole, the one with the smallest priority.
func covering(spans []Span, from, to time.Time) (Stage, bool) {
	var best Stage
	found := false
	for _, s := range spans {
		if s.Start.After(from) || s.End.Before(to) {
			continue
		}
		if !found || s.Stage.priority() < best.priority() {
			best, found = s.Stage, true
		}
	}
	return best, found
}

// AsleepDuration is the length of the time ACTUALLY spent asleep in the spans:
// the union of the sleep stages, so an overlap counts once, `awake` beats the
// sleep stage under it, and `inBed` counts only where nothing more specific
// covers it — that is, not at all.
func AsleepDuration(spans []Span) time.Duration {
	slices, _ := Resolve(spans)
	return asleep(slices)
}

func asleep(slices []Span) time.Duration {
	var total time.Duration
	for _, s := range slices {
		if s.Stage.IsAsleep() {
			total += s.Duration()
		}
	}
	return total
}

// NightGap: a break longer than this starts a new sleep session.
//
// Three hours: long enough that waking in the night does not cut the night in
// two, short enough that an afternoon nap ends up on its own. The same value
// lives in the phone's `SleepAnalysisBuilder.nightGapSeconds` and in the web's
// `lib/sleep.ts` — deliberately, so that all three say the same night.
const NightGap = 3 * time.Hour

// Night is one sleep session, already flattened.
type Night struct {
	// Slices are disjoint and ordered by time (`inBed` among them, wherever no
	// more specific stage covers it).
	Slices []Span
	// Overlap is how much the raw stages of this night covered each other.
	Overlap time.Duration
}

// Asleep is the time actually spent asleep.
func (n Night) Asleep() time.Duration { return asleep(n.Slices) }

// Onset is the start of the first SLEEP slice, Wake the end of the last one.
// `ok` is false for a session that holds no sleep at all (only awake / in bed):
// that night has no falling-asleep moment, and inventing one would be worse than
// leaving the timing rules without it.
func (n Night) Onset() (time.Time, bool) {
	for _, s := range n.Slices {
		if s.Stage.IsAsleep() {
			return s.Start, true
		}
	}
	return time.Time{}, false
}

func (n Night) Wake() (time.Time, bool) {
	for i := len(n.Slices) - 1; i >= 0; i-- {
		if n.Slices[i].Stage.IsAsleep() {
			return n.Slices[i].End, true
		}
	}
	return time.Time{}, false
}

// Nights threads the spans into sessions and flattens each one.
//
// The gap is measured against the RUNNING MAXIMUM end, not against the end of
// the previous span: `inBed` overlaps the whole night, so the stages after it
// start "earlier" than where the previous one ended.
func Nights(spans []Span, gap time.Duration) []Night {
	groups := group(spans, gap)
	out := make([]Night, 0, len(groups))
	for _, g := range groups {
		slices, overlap := Resolve(g)
		out = append(out, Night{Slices: slices, Overlap: overlap})
	}
	return out
}

// group splits the (start-ordered) spans into sessions.
func group(spans []Span, gap time.Duration) [][]Span {
	ordered := make([]Span, len(spans))
	copy(ordered, spans)
	sort.Slice(ordered, func(i, j int) bool { return ordered[i].Start.Before(ordered[j].Start) })

	var out [][]Span
	var current []Span
	var runningEnd time.Time
	for _, s := range ordered {
		if len(current) > 0 && s.Start.Sub(runningEnd) > gap {
			out = append(out, current)
			current = nil
			runningEnd = time.Time{}
		}
		current = append(current, s)
		if s.End.After(runningEnd) {
			runningEnd = s.End
		}
	}
	if len(current) > 0 {
		out = append(out, current)
	}
	return out
}
